### DISCUSSION TITLE
Leading stars appear after updating org-modern

### DISCUSSION THREAD

**Discussion by mok0** [minad/org-modern#227] [Q&A]
After updating `org-modern` from the git repo, I found that my bullets was suddenly replaced by triangles. I was able to figure out that I can get the original appearance, which is what I want, by setting
```
(setq org-modern-star 'replace)
```

in my init file. However, now  leading asterisks appears on every headline except level 1, which it didn't before. How do I get rid of it?

**Comment by mok0** [comment]
I found what the problem is... sort of. I know it has to do with `ef-themes`, but I don't know what `ef-themes` is doing that results in this behaviour. I disabled it and went back to my old theme and the correct behaviour was restored, so this problem has nothing to do with `org-modern`. 

Therefore I close this discussion.

**Reply by refaelsh** [reply]
> I found what the problem is... sort of. I know it has to do with `ef-themes`, but I don't know what `ef-themes` is doing that results in this behaviour. I disabled it and went back to my old theme and the correct behaviour was restored, so this problem has nothing to do with `org-modern`.
> 
> Therefore I close this discussion.

I am using the doom-one theme. I tried removing it from my init.el and the problem still remains.

**Comment by Gleek** [comment]
This started to happen with me recently.
I drilled down the issue to when:
- org-indent-mode is enabled
- Following ligature is enabled.
 
```elisp
(let ((alist '((42 . ".\\(?:\\(?:\\*\\*/\\)\\|\\(?:\\*[*/]\\)\\|[*/>]\\)"))))
  (dolist (char-regexp alist)
    (set-char-table-range composition-function-table (car char-regexp)
                          `([,(cdr char-regexp) 0 font-shape-gstring]))))
```
Removed this ligature rule to get it to work correctly.

This did not happen before, and started to happen with the recent update. Might be a regression in org-modern itself.

For reference this is how it looked:
<img width="406" alt="image" src="https://github.com/user-attachments/assets/8d2e6f9b-c72f-4e55-a4a9-9fd8bfc38637" />

**Comment by refaelsh** [comment]
> After updating `org-modern` from the git repo, I found that my bullets was suddenly replaced by triangles. I was able to figure out that I can get the original appearance, which is what I want, by setting
> 
> ```
> (setq org-modern-star 'replace)
> ```
> 
> in my init file. However, now leading asterisks appears on every headline except level 1, which it didn't before. How do I get rid of it?

I am having the exact same problem.

**Comment by mbc90** [comment]
Adding to confirm Gleek's problem that this a ligature interaction.

In my case I'm using [`ligature.el`](https://github.com/mickeynp/ligature.el) 
with `(global-ligature-mode t)` and an asterisk-containing pattern (`"***" and "**"`) 
in my ligature list. Same symptom: leading stars visible on level 2+ 
headings despite  `org-modern-hide-stars t`, 
and `org-hide`/`org-indent` faces correctly inheriting a foreground that 
matches the background.

Highlighted here are the items causing the issue:
<img width="1135" height="484" alt="trouble-line" src="https://github.com/user-attachments/assets/8df54e41-01d4-409a-a7e8-eb4fd79b5ee5" />

### How to verify it's this issue

Put cursor on the visible `*` after the bullet and run `C-u C-x =`. 
You'll see the actual character is a SPACE, with text properties:
```
line-prefix    #("*" 0 1 (face org-indent))
wrap-prefix    #("*** " 0 1 (face org-indent) 1 4 (face org-indent))
```
Then `M-: (face-foreground 'org-indent nil t)` returns your background color 
correctly confirming the face *should* be invisible. But the ligature 
engine renders the `***` and `**` glyph at a layer that bypasses the face foreground, 
so the stars become visible anyway. 
I am using doom-themes in my config, however I found this issue was present with the modus-operandi themes. 
To recreate, copy these two source blocks

save as `modern.el`:
```elisp
;;; Reproduction for org-modern + ligature.el leading-stars issue
;;; Run with: emacs -Q -l modern.el path/to/test.org

(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)

;; Refresh archives if needed
(unless package-archive-contents
  (package-refresh-contents))

;; Install required packages
(package-install 'org-modern)
(package-install 'ligature)
;; Load packages
(require 'org-modern)
(require 'ligature)

;; Minimal UI
(menu-bar-mode -1)
(tool-bar-mode -1)
(when (display-graphic-p)
  (scroll-bar-mode -1))

;; Use built-in modus-operandi (Emacs 28+)
(load-theme 'modus-operandi t)

;; Org settings
(setq org-auto-align-tags nil
      org-tags-column 0
      org-catch-invisible-edits 'show-and-error
      org-special-ctrl-a/e t
      org-insert-heading-respect-content t
      org-hide-emphasis-markers t
      org-pretty-entities t
      org-agenda-tags-column 0
      org-ellipsis "…"
      org-startup-indented t)

;; Enable org-modern
(global-org-modern-mode)

;; Enable ligatures with the asterisk patterns that trigger the bug
(ligature-set-ligatures 't '("***" "**" "*>" "*/" "->" "=>" "!=" "==" "===" "<=>" "<-" "<--"))
(global-ligature-mode t)

```
                      
Save as `test.org`:

```org
* Level One Heading
Some content here.
** Level Two Heading
The leading star *should* be invisible but renders as visible asterisk.
*** Level Three Heading
Same problem here.
**** Level Four Heading
asterisks are now gone.
```
And then run:
```bash
emacs -Q -l modern.el test.org
```
Turn ligature mode off to confirm: `M-x ligature-mode` to toggle it off. The asterisks will disappear

### Workaround

Either disable ligatures in `org-mode`, or remove asterisk-containing 
patterns from your ligature list:
```elisp
(ligature-set-ligatures 'prog-mode '(...))   ;; instead of 't
```
Toggling `(global-ligature-mode -1)` is a quick way to confirm before 
making the config change permanent.

Personally, I  just removed `"***"` and `"**"` from the list due to not using them.
### Note on root cause

This doesn't seem to be an org-modern bug `org-indent-mode` is what 
adds the `line-prefix` stars, and ligature substitution bypassing face 
attributes is an Emacs rendering behavior. But since the symptom 
shows up specifically when I switched to org-modern  from [org-bullets](https://github.com/sabof/org-bullets) (I can confirm this does not happen with org-bullets), a brief note in the  README under known Incompatibilities might save the next person a long 
debugging session.
